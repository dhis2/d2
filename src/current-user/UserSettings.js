import Api from '../api/Api';
import { isString } from '../lib/check';


/**
 * @description
 * Handles communication with the userSettings endpoint. Can be used to get or save userSettings.
 *
 * @memberof module:current-user
 */

class UserSettings {
    constructor(userSettings, api = Api.getApi()) {
        this.api = api;

        if (userSettings) {
            this.settings = userSettings;
        }
    }

    /**
     * @returns {Promise} Promise that resolves with the usersettings object from the api.
     *
     * @description
     * Loads all the user settings of current user and returns them as an object from the promise.
     * ```js
     * d2.currentUser.userSettings.all()
     *  .then(userSettings => {
     *    console.log('UI Locale: ' + userSettings.keyUiLocale);
     *  });
     * ```
     */
    all() {
        return this.settings
            ? Promise.resolve(this.settings)
            : this.api.get('userSettings')
                .then((userSettings) => { this.settings = userSettings; return Promise.resolve(this.settings); });
    }

    /**
     * @param {String} key The identifier of the user setting that should be retrieved.
     * @returns {Promise} A promise that resolves with the value or will fail if the value is not available.
     *
     * @description
     * ```js
     * d2.currentUser.userSettings.get('keyUiLocale')
     *  .then(userSettingValue => {
     *    console.log('UI Locale: ' + userSettingValue);
     *  });
     * ```
     */
    get(key) {
        if (this.settings && this.settings[key]) {
            return Promise.resolve(this.settings[key]);
        }

        function processValue(value) {
            // Attempt to parse the response as JSON. If this fails we return the value as is.
            try {
                return JSON.parse(value);
            } catch (e) {
                return value;
            }
        }

        return new Promise((resolve, reject) => {
            if (!isString(key)) {
                throw new TypeError('A "key" parameter should be specified when calling get() on userSettings');
            }

            this.api.get(['userSettings', key].join('/'))
                .then((response) => {
                    const value = processValue(response);
                    // Store the value on the user settings object
                    this[key] = value;
                    if (value) {
                        resolve(value);
                    }
                    reject(new Error('The requested userSetting has no value or does not exist.'));
                });
        });
    }

    /**
     * @param {String} key The identifier of the user setting that should be saved.
     * @param {String} value The new value of the user setting.
     * @returns {Promise} A promise that will resolve when the new value has been saved, or fail if saving fails.
     *
     * @description
     * ```js
     * d2.currentUser.userSettings.set('keyUiLocale', 'fr')
     *  .then(() => {
     *   console.log('UI Locale is now "fr");
     *  });
     * ```
     */
    set(key, value) {
        delete this.settings;

        const settingUrl = ['userSettings', key].join('/');
        if (value === null || (`${value}`).length === 0) {
            return this.api.delete(settingUrl);
        }
        return this.api.post(settingUrl, value, { headers: { 'Content-Type': 'text/plain' } });
    }
}

export default UserSettings;
