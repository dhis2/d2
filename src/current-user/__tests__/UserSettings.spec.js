import MockApi from '../../api/Api'
import UserSettings from '../UserSettings'

jest.mock('../../../src/api/Api')

describe('CurrentUser.userSettings', () => {
    const userSettingsFixture = { keyUiLocale: 'en' }
    let userSettings

    beforeEach(() => {
        userSettings = new UserSettings()
    })

    afterEach(() => {
        MockApi.mockReset()
    })

    it('should not be allowed to be called without new', () => {
        expect(() => UserSettings()).toThrowErrorMatchingSnapshot()
    })

    it('should set an instance of MockApi onto the UserSettings instance', () => {
        expect(userSettings.api).toBe(MockApi.getApi())
    })

    describe('all', () => {
        beforeEach(() => {
            userSettings.api.get.mockReturnValue(
                Promise.resolve(userSettingsFixture)
            )
        })

        it('should be a function', () => {
            expect(userSettings.all).toBeInstanceOf(Function)
        })

        it('should call the api to get all the userSettings', () => {
            expect.assertions(2)

            return userSettings.all().then(() => {
                expect(userSettings.api.get).toHaveBeenCalledTimes(1)
                expect(userSettings.api.get.mock.calls[0][0]).toBe(
                    'userSettings'
                )
            })
        })

        it('should resolve the promise with the settings', () => {
            expect.assertions(1)

            return userSettings.all().then(settings => {
                expect(settings.keyUiLocale).toBe('en')
            })
        })

        it('should cache the current user settings', () => {
            expect.assertions(2)

            return userSettings
                .all()
                .then(() => userSettings.all())
                .then(() => {
                    expect(userSettings.api.get).toHaveBeenCalledTimes(1)
                    expect(userSettings.settings).toEqual(userSettingsFixture)
                })
        })
    })

    describe('get', () => {
        beforeEach(() => {
            userSettings.api.get.mockReturnValue(
                Promise.resolve(userSettingsFixture.keyUiLocale)
            )
        })

        it('should be a function', () => {
            expect(userSettings.get).toBeInstanceOf(Function)
        })

        it('should return a Promise', () => {
            const result = userSettings.get('keyUiLocale')

            expect(result).toBeInstanceOf(Promise)
        })

        it('should reject the promise with an error if no key has been specified', () => {
            expect.assertions(2)

            return userSettings.get().catch(error => {
                expect(error).toBeInstanceOf(TypeError)
                expect(error.message).toBe(
                    'A "key" parameter should be specified when calling get() on userSettings'
                )
            })
        })

        it('should call the api to get the value', () => {
            userSettings.get('keyUiLocale')

            expect(userSettings.api.get).toBeCalledWith(
                'userSettings/keyUiLocale'
            )
        })

        it('should return the value from the promise', () => {
            expect.assertions(1)

            return userSettings.get('keyUiLocale').then(value => {
                expect(value).toBe('en')
            })
        })

        it('should try to transform the response to json if possible', () => {
            userSettings.api.get.mockReturnValueOnce(
                Promise.resolve('{"mydataKey": "myDataValue"}')
            )

            expect.assertions(1)

            return userSettings.get('keyUiLocale').then(value => {
                expect(value).toEqual({ mydataKey: 'myDataValue' })
            })
        })

        it('should reject the promise if the value is empty', () => {
            userSettings.api.get.mockReturnValueOnce(Promise.resolve(''))

            expect.assertions(1)

            return userSettings
                .get('keyThatDefinitelyDoesNotExist')
                .catch(error => {
                    expect(error.message).toBe(
                        'The requested userSetting has no value or does not exist.'
                    )
                })
        })

        it('should use the cache', () => {
            userSettings.api.get.mockReturnValueOnce(
                Promise.resolve(userSettingsFixture)
            )

            expect.assertions(2)

            return userSettings
                .all()
                .then(() => userSettings.get('keyUiLocale'))
                .then(value => {
                    expect(userSettings.api.get).toHaveBeenCalledTimes(1)
                    expect(value).toBe(userSettingsFixture.keyUiLocale)
                })
        })

        it('should also return a promise when serving cached values', () => {
            userSettings.api.get.mockReturnValueOnce(
                Promise.resolve(userSettingsFixture)
            )

            expect.assertions(1)

            return userSettings.all().then(() => {
                expect(userSettings.get('keyUiLocale')).toBeInstanceOf(Promise)
            })
        })
    })

    describe('set', () => {
        beforeEach(() => {
            userSettings.api.get.mockReturnValue(
                Promise.resolve(userSettingsFixture)
            )
            userSettings.api.post.mockReturnValueOnce(Promise.resolve())
            userSettings.api.delete.mockReturnValueOnce(Promise.resolve())
        })

        afterEach(() => {
            userSettings = new UserSettings()
        })

        it('should POST to the API', () =>
            userSettings.set('mySetting', 'my value').then(() => {
                expect(userSettings.api.get).toHaveBeenCalledTimes(0)
                expect(userSettings.api.post).toHaveBeenCalledTimes(1)
                expect(userSettings.api.delete).toHaveBeenCalledTimes(0)
            }))

        it('should DELETE if the value is null or an empty string', () =>
            userSettings.set('mySetting', '').then(() => {
                expect(userSettings.api.get).toHaveBeenCalledTimes(0)
                expect(userSettings.api.post).toHaveBeenCalledTimes(0)
                expect(userSettings.api.delete).toHaveBeenCalledTimes(1)
            }))

        it('should clear out the cache', () =>
            userSettings
                .all()
                .then(() => userSettings.all())
                .then(() => userSettings.set('a', 'b'))
                .then(() => userSettings.all())
                .then(() => userSettings.all())
                .then(() => {
                    expect(userSettings.api.post).toHaveBeenCalledTimes(1)
                    expect(userSettings.api.get).toHaveBeenCalledTimes(2)
                }))
    })
})
