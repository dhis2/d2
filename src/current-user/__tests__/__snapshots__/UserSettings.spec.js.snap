// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentUser.userSettings should not be allowed to be called without new 1`] = `"Class constructor UserSettings cannot be invoked without 'new'"`;
